lib/services/fanout.go

```diff
@@ -0,0 +1,255 @@
+				remove = append(remove, entry.watcher)
+				continue Inner
+			}
+			if !match {
+				continue Inner
+			}
+			emitEvent := event
+			// if this entry loads secrets, emit the
+			// full unfiltered event.
+			if entry.kind.LoadSecrets {
+				emitEvent = fullEvent
+			}
+			if err := entry.watcher.emit(emitEvent); err != nil {
+				remove = append(remove, entry.watcher)
+				continue Inner
+			}
+		}
+		for _, w := range remove {
+			f.removeWatcher(w)
+			w.cancel()
+		}
+	}
+}
+
+// CloseWatchers closes all attached watchers, effectively
+// resetting the Fanout instance.
+func (f *Fanout) CloseWatchers() {
+	f.Lock()
+	defer f.Unlock()
+	for kind, entries := range f.watchers {
+		for _, entry := range entries {
+			entry.watcher.cancel()
+		}
+		delete(f.watchers, kind)
+	}
+	// watcher map was potentially quite large, so
+	// relenguish that memory.
+	f.watchers = make(map[string][]fanoutEntry)
+}
+
+func (f *Fanout) addWatcher(w *fanoutWatcher) {
+	for _, kind := range w.watch.Kinds {
+		entries := f.watchers[kind.Kind]
+		entries = append(entries, fanoutEntry{
+			kind:    kind,
+			watcher: w,
+		})
+		f.watchers[kind.Kind] = entries
+	}
+}
+
+func (f *Fanout) removeWatcher(w *fanoutWatcher) bool {
+	var found bool
+	for _, kind := range w.watch.Kinds {
+		entries := f.watchers[kind.Kind]
+	Inner:
+		for i, entry := range entries {
+			if entry.watcher == w {
+				found = true
+				entries = append(entries[:i], entries[i+1:]...)
+				break Inner
+			}
+		}
+		switch len(entries) {
+		case 0:
+			delete(f.watchers, kind.Kind)
+		default:
+			f.watchers[kind.Kind] = entries
+		}
+	}
+	return found
+}
+
+func newFanoutWatcher(ctx context.Context, watch Watch) (*fanoutWatcher, error) {
+	if len(watch.Kinds) < 1 {
+		return nil, trace.BadParameter("must specify at least one resource kind to watch")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	if watch.QueueSize < 1 {
+		watch.QueueSize = defaultQueueSize
+	}
+	return &fanoutWatcher{
+		watch:  watch,
+		eventC: make(chan Event, watch.QueueSize),
+		cancel: cancel,
+		ctx:    ctx,
+	}, nil
+}
+
+type fanoutWatcher struct {
+	emux   sync.Mutex
+	err    error
+	watch  Watch
+	eventC chan Event
+	cancel context.CancelFunc
+	ctx    context.Context
+}
+
+func (w *fanoutWatcher) emit(event Event) error {
+	select {
+	case <-w.ctx.Done():
+		return trace.Wrap(w.ctx.Err(), "watcher closed")
+	case w.eventC <- event:
+		return nil
+	default:
+		return trace.BadParameter("buffer overflow")
+	}
+}
+
+func (w *fanoutWatcher) Events() <-chan Event {
+	return w.eventC
+}
+
+func (w *fanoutWatcher) Done() <-chan struct{} {
+	return w.ctx.Done()
+}
+
+func (w *fanoutWatcher) Close() error {
+	w.cancel()
+	return nil
+}
+
+func (w *fanoutWatcher) setError(err error) {
+	w.emux.Lock()
+	defer w.emux.Unlock()
+	w.err = err
+}
+
+func (w *fanoutWatcher) getError() error {
+	w.emux.Lock()
+	defer w.emux.Unlock()
+	return w.err
+}
+
+func (w *fanoutWatcher) Error() error {
+	return w.getError()
+}
```

Comment:
Consider merging these two

Reply:
Also, I can't see where this error is checked

Reply:
`Error` is part of `Watcher` interface and is checked when `fanoutWatcher` is being used in that capacity (more specifically, `Watcher`s returned by `Cache.NewWatcher` end up using `fanoutWatcher` as their underlying implementation).